Hide orphaned entities

dperelman opened the pull request:
Search through `data.raw` to find everything that was referenced before Pacifist removed whatever it removes and is no longer referenced after and remove all of those things as well. This effectively removes things like corpses/remnants of removed entities and particles/explosions related to weapons/combat robots that get removed. This PR also adds some new dummies to cover some categories that now end up empty. (Note the dummies change may conflict with #57, but they should be straight-forward to merge cleanly as the changes are actually either identical or just next to each other, not overlapping.)

Note the way this checks for references just looks for any `string`s matching the names of entries in `data.raw`, which technically is an overapproximation, meaning that it may think something is referenced when it isn't, which would just result in that thing actually being left in (and therefore visible in the rich text icon selection GUI) accidentally. This is unlikely to happen in practice. The "right" solution would be to determine which `string`s are `EntityID`s or whatever `*ID`s and only treat those as references, but that seems like overkill here. (I might write the code to do that anyway since someone asked for it in the Factorio Discord when I asked about this, but would probably not bother using it for this project.)

After this pass, there are some explosions left for friendly buildings getting damaged and the atomic bomb explosion for the reactor being destroyed. Those explosions started showing up in the rich text icon selection GUI, so I added a couple lines to `PacifistMod.remove_misc()` to hide them. I then confirmed that destroying a reactor still produces an atomic bomb explosion as expected.

Together with #57, this should fix #56.

dperelman commented on functions.lua:
```diff
@@ -241,6 +241,72 @@ function PacifistMod.remove_misc()
+function PacifistMod.remove_orphaned_entities(was_used)
+    repeat
+        local changed = false
+        local is_used = PacifistMod.enumerate_used_names()
```
Hmm... I haven't tried it, I can test with SE/Seablock/other overhauls. But I think I can rework the algorithm to only do the full read of `data.raw` once and record which things reference which names instead of just whether a name is referenced. If we also had a list of everything Pacifist removes (which shouldn't be too hard to maintain...), then we can avoid ever doing any more iterations of even the list of names in `data.raw` (at that point, it's basically a reference counting garbage collector). Which is only moderately more complicated and should be notably faster.

Derpumu replied:
SE isn't compatible, but Seablock should be, as well as Exotic Industries - those are so far the largest mod packs I'd like to support

dperelman replied:
FYI, SeaBlock fails to load. I added it as an optional dependency because it was erroring on something added after Pacifist ran, but now it says

> Technology military: Non-contiguous levels: 1, followed by 3

and I did not investigate further.

dperelman replied:
I have a draft of the new algorithm in a branch https://github.com/dperelman/Pacifist/tree/feature/remove-orphaned-entities-opt . I'll merge that into this PR after I do a bit more clean up and comment writing.

Here's my performance measurements. Numbers are the timestamp on the log line that says "Factorio initialised", which I believe is seconds since Factorio started. I ran each configuration a few times and took the last one to avoid caching effects (the numbers usually dropped slightly after the first run). The left column is with just the base mod, the right column is with "Exotic Industries" loaded (i.e. just the default of whatever that mod depends on, no optional dependencies). Each row is a different configuration of Pacifist. The third row is this PR as it currently is, the final row is the branch linked at the start of this comment as it currently is.

Surprisingly, Pacifist makes the game load significantly *faster*, I guess because it's removing things, so later phases of the loading process have less work to do?
| mod           | base    | EI     |
|:--------------|--------:|-------:|
| no Pacifist   |  13.181 | 19.802 |
| `main`        |  11.404 | 16.909 |
| naive orphans |  10.897 | 17.563 |
| opt orphans   |   9.617 | 14.936 |

Derpumu commented:
I tried to resolve the conflicts after mergin the other PR, but now I get this and am not quite sure what the reason is:
![image](https://github.com/Derpumu/Pacifist/assets/183507/3cc1121a-facd-4cca-87e5-8bda1a7d3331)

dperelman commented:
I couldn't reproduce that error message. But when investigating I noticed that `damage-type` references are in fields of name `type` so they weren't getting handled properly anyway since I was ignoring all fields of name `type` because they're usually the `type` of that object, not a reference to something else. So I changed the code to never delete `damage-type`s. [DamageType](https://lua-api.factorio.com/latest/prototypes/DamageType.html) can't reference other things anyway, so this won't result in keeping other objects around.

At the same time, it occurred to me that this code shouldn't delete anything labeled as an exception due to the mod rules, so I added a check for that as well. Not sure it does anything, but it would be confusing if it accidentally overrode the exceptions.

Derpumu commented:
I'll have a look at it again in the next days.

dperelman commented:
Okay, sounds like this strategy doesn't work in general as we can't tell which entities mods will reference by name in their own code. The options I see are:

1. Abandon this.
2. Hide instead of remove. (I just pushed this version)
3. For each mod with compatibility issues, hard code a list of entities that have references (these would be added to the `exceptions` collection).

(3) sounds brittle and like a lot of work for minor gain, so I'd lean towards (2). Noting that if it accidentally hides extra things, we can add them to the exceptions, but I doubt that's likely to be a problem.

One problem with (2) is that it doesn't hide combat robots, so I want to extend it with an option to "pretend delete" entities, so it will hide combat robots and everything they reference since removing them and everything they reference isn't working for all mods.

(I saw that SeaBlock error before on a different branch and fixed it by marking SeaBlock as a dependency so it would load before Pacifist... but then I got a different error, which I why I never pushed that change.)

dperelman commented:
With these changes, I got a similar error on Exotic Industries about `stone-wall` (that is, once I made sure `destroyer-remnants` wasn't being removed. Adding it to `mods_require_walls` fixed it, but I'm not sure what's going on if you had tested it before and not needed that.

Derpumu commented:
I'll run it against my test save games later this week when I'm back home. I am thinking about logging which types of things get hidden by this and maybe restrict the algorithm to things like remnants etc that actually show up in the signals and RTF icons. 

The defender remnants maybe should be added to the exceptions for Exotic Industries either way, but I'll look into that later as well. 

Could you tell me how to reproduce the stone-wall issue with EI? I've been running a save with EI and Pacifist for a while without issues. Maybe you have different mods or mod settings?

In any case, thank you again for the work!

arnemertz commented on functions.lua:
```diff
@@ -250,6 +254,166 @@ function PacifistMod.remove_misc()
+        end
+    end
+
+    -- Helper that recurses through an object x which is data.row[from_group][from_name]
```
```suggestion
    -- Helper that recurses through an object x which is data.raw[from_group][from_name]
```

arnemertz commented on lib/data_raw.lua:
```diff
@@ -1,17 +1,42 @@
 local data_raw = {}
+data_raw.removed = {}
```
from what I gather this is both for removed and hidden things. Can this lead to problems or is the distinction unnecessary?

```suggestion
data_raw.removed_or_hidden = {}
```

Derpumu commented on functions.lua:
```diff
@@ -250,6 +254,166 @@ function PacifistMod.remove_misc()
+        end
+    end
+
+    -- Helper that recurses through an object x which is data.row[from_group][from_name]
```
```suggestion
    -- Helper that recurses through an object x which is data.raw[from_group][from_name]
```

dperelman replied:
Oops. I never actually pushed the fix for this typo, so it ended up in `main`.

Derpumu replied:
I'll fix it next time I work on it (if I remember)

dperelman replied:
ca3f45d

Derpumu commented on lib/data_raw.lua:
```diff
@@ -1,17 +1,42 @@
 local data_raw = {}
+data_raw.removed = {}
```
should we rename this to `removed_or_hidden`?

dperelman replied:
Hmm... that implies that it gets everything `.hide()` is called on... which might actually be the right thing to do anyway.
